log4j-api/src/main/java/org/apache/logging/log4j/spi/internal/DefaultScopedContextProvider.java

```diff
@@ -0,0 +1,399 @@
+            return addObject(key, value);
+        }
+
+        /**
+         * Adds a key/value pair to the ScopedContext being constructed.
+         *
+         * @param key      the key to add.
+         * @param supplier the function to generate the value.
+         * @return the ScopedContext being constructed.
+         */
+        @Override
+        public Instance where(final String key, final Supplier<Object> supplier) {
+            return addObject(key, supplier.get());
+        }
+
+        private Instance addObject(final String key, final Object obj) {
+            return obj != null ? new Instance(this, key, obj) : this;
+        }
+
+        /**
+         * Executes a code block that includes all the key/value pairs added to the ScopedContext.
+         *
+         * @param task the code block to execute.
+         */
+        @Override
+        public void run(final Runnable task) {
+            new Runner(this, null, null, task).run();
+        }
+
+        /**
+         * Executes a code block that includes all the key/value pairs added to the ScopedContext on a different Thread.
+         *
+         * @param task the code block to execute.
+         * @return a Future representing pending completion of the task
+         */
+        @Override
+        public Future<Void> run(final ExecutorService executorService, final Runnable task) {
+            return executorService.submit(
+                    new Runner(this, ThreadContext.getContext(), ThreadContext.getImmutableStack(), task), null);
+        }
+
+        /**
+         * Executes a code block that includes all the key/value pairs added to the ScopedContext.
+         *
+         * @param task the code block to execute.
+         * @return the return value from the code block.
+         */
+        @Override
+        public <R> R call(final Callable<R> task) throws Exception {
+            return new Caller<>(this, null, null, task).call();
+        }
+
+        /**
+         * Executes a code block that includes all the key/value pairs added to the ScopedContext on a different Thread.
+         *
+         * @param task the code block to execute.
+         * @return a Future representing pending completion of the task
+         */
+        @Override
+        public <R> Future<R> call(final ExecutorService executorService, final Callable<R> task) {
+            return executorService.submit(
+                    new Caller<>(this, ThreadContext.getContext(), ThreadContext.getImmutableStack(), task));
+        }
+
+        private DefaultScopedContextProvider getProvider() {
+            return provider;
+        }
+    }
+
+    private static class Runner implements Runnable {
+        private final Map<String, Object> contextMap = new HashMap<>();
+        private final Map<String, String> threadContextMap;
+        private final ThreadContext.ContextStack contextStack;
+        private final Instance context;
```

Comment:
Should we integrate the propagation of `ThreadContext` into `ScopedContext`?

I have mixed feelings about that. Since this API is supposed to replace `ThreadContext`, maybe we should keep the propagation of the two separate. Users of `ThreadContext` do have ways to propagate it anyway.

Reply:
I believe we should leave it integrated. Even if we deprecate ThreadContext (which you said you reconsidered due to SLF4J) it would be years before it could be removed. In the meantime it is better to support them both until we do remove ThreadContext.

Reply:
The contract of `ScopedContext` is that it propagates to other threads in a modern way, we don't need to promise anything about `ThreadContext`.

If users are not satisfied about how `ThreadContext` fails to propagate, they can always migrate to `ScopedContext`. :wink:

Reply:
You say that is if they won't have code that already has ThreadContext stuff while they are introducing ScopedContext. Nobody changes everything at once unless they are starting from scratch.  If users abandon ThreadContext then storing it essentially does nothing.

Again, I don't see a problem.

Reply:
If they have legacy code, they already propagate `ThreadContext` themselves… no need to do it for them.

Imagine I want to provide a custom implementation of `ScopedContextProvider`. Do I need to propagate `ThreadContext` too?

Reply:
Propagation of the ThreadContext is now not automatic. It requires a call to .withThreadContext(true) to be enabled.